File: visual-studio/SecretLoader/Vault/SecretFetcher.cs
/*
 * SecretLoader
 * Copyright (C) 2026 Kivi A.Ş.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace SecretLoader
{
    internal sealed class FetchException : Exception
    {
        public FetchException(string message) : base(message) { }
    }

    /// <summary>A project returned by the project-list command: an id to inject and a human label to show.</summary>
    internal sealed class ProjectEntry
    {
        public string Id { get; }
        public string Label { get; }
        public ProjectEntry(string id, string label) { Id = id; Label = label; }
    }

    /// <summary>
    /// Runs the configured vault CLI and parses its JSON output into key/value secrets. C# port of the
    /// JetBrains SecretFetcher: quote-aware tokenization, argument-injection validation, drop {project}
    /// when blank (CLI self-resolves), timeout, and jsonPath extraction.
    /// </summary>
    internal static class SecretFetcher
    {
        // Reject whitespace and anything that could inject extra CLI arguments.
        private static readonly Regex SafeArg = new Regex("^[A-Za-z0-9_.:/@-]+$", RegexOptions.Compiled);

        public static async Task<IReadOnlyDictionary<string, string>> FetchAsync(
            SecretLoaderSettings settings, string env, string projectId, string workingDir, CancellationToken ct)
        {
            List<string> command = BuildCommand(settings.CliPath, settings.CommandTemplate, env, projectId);
            if (command.Count == 0) throw new FetchException("Command is empty after substitution.");

            string stdout = await RunAsync(command, settings.CliPath, workingDir, settings.TimeoutSeconds, ct).ConfigureAwait(false);
            return Parse(stdout, settings.JsonPath);
        }

        /// <summary>Runs the configured project-list command and parses its JSON into id/label entries (for the picker).</summary>
        public static async Task<IReadOnlyList<ProjectEntry>> ListProjectsAsync(
            SecretLoaderSettings settings, string env, string workingDir, CancellationToken ct)
        {
            List<string> command = BuildListCommand(settings.CliPath, settings.ListProjectsCommand, env);
            if (command.Count == 0) throw new FetchException("No project-list command is configured.");

            string stdout = await RunAsync(command, settings.CliPath, workingDir, settings.TimeoutSeconds, ct).ConfigureAwait(false);
            return ParseProjects(stdout);
        }

        /// <summary>Starts the CLI, enforces the timeout, checks the exit code, and returns stdout.</summary>
        private static async Task<string> RunAsync(List<string> command, string cliPath, string workingDir, int timeoutSeconds, CancellationToken ct)
        {
            var psi = new ProcessStartInfo
            {
                FileName = command[0],
                Arguments = string.Join(" ", command.Skip(1).Select(QuoteArg)),
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                WorkingDirectory = workingDir,
            };

            Process process;
            try
            {
                process = Process.Start(psi) ?? throw new FetchException("Failed to start CLI process.");
            }
            catch (Exception e) when (e is not FetchException)
            {
                throw new FetchException($"Could not start '{cliPath}': {e.Message}");
            }

            using (process)
            {
                Task<string> outTask = process.StandardOutput.ReadToEndAsync();
                Task<string> errTask = process.StandardError.ReadToEndAsync();
                Task exitTask = Task.Run(() => process.WaitForExit());

                Task finished = await Task.WhenAny(exitTask, Task.Delay(timeoutSeconds * 1000, ct)).ConfigureAwait(false);
                if (finished != exitTask)
                {
                    try { process.Kill(); } catch { /* ignore */ }
                    throw new FetchException($"CLI timed out after {timeoutSeconds}s.");
                }

                string stdout = await outTask.ConfigureAwait(false);
                string stderr = await errTask.ConfigureAwait(false);

                if (process.ExitCode != 0)
                {
                    string err = stderr.Trim();
                    if (err.Length > 500) err = err.Substring(0, 500);
                    throw new FetchException($"CLI exited {process.ExitCode}: {err}");
                }
                return stdout;
            }
        }

        /// <summary>Tokenizes the project-list template, substituting {cli} and {env} (no {project}).</summary>
        public static List<string> BuildListCommand(string cliPath, string template, string env)
        {
            if (string.IsNullOrWhiteSpace(template)) return new List<string>();
            if (env.Length > 0 && !SafeArg.IsMatch(env))
                throw new FetchException($"Unsafe environment name '{env}' (only [A-Za-z0-9_.:/@-] allowed).");

            var outList = new List<string>();
            foreach (var token in Tokenize(template))
            {
                outList.Add(token == "{cli}" ? cliPath : token.Replace("{env}", env));
            }
            return outList.Where(t => !string.IsNullOrWhiteSpace(t)).ToList();
        }

        /// <summary>Parses a project-list JSON: array of strings, or array of objects (id/name/slug…), under root or .projects/.workspaces/.data.</summary>
        public static List<ProjectEntry> ParseProjects(string output)
        {
            var list = new List<ProjectEntry>();
            if (string.IsNullOrWhiteSpace(output)) return list;

            JToken node;
            try { node = JToken.Parse(output); }
            catch (Exception e) { throw new FetchException($"Failed to parse project-list JSON: {e.Message}"); }

            JArray? arr = node as JArray
                          ?? node["projects"] as JArray
                          ?? node["workspaces"] as JArray
                          ?? node["data"] as JArray;
            if (arr == null) return list;

            foreach (var item in arr)
            {
                string? id, label;
                if (item is JValue)
                {
                    id = label = (string?)item;
                }
                else
                {
                    id = (string?)(item["id"] ?? item["projectId"] ?? item["workspaceId"] ?? item["slug"] ?? item["_id"]);
                    label = (string?)(item["name"] ?? item["projectName"] ?? item["slug"]) ?? id;
                }
                if (!string.IsNullOrEmpty(id)) list.Add(new ProjectEntry(id!, label ?? id!));
            }
            return list;
        }

        public static List<string> BuildCommand(string cliPath, string template, string env, string projectId)
        {
            if (env.Length > 0 && !SafeArg.IsMatch(env))
                throw new FetchException($"Unsafe environment name '{env}' (only [A-Za-z0-9_.:/@-] allowed).");
            if (projectId.Length > 0 && !SafeArg.IsMatch(projectId))
                throw new FetchException("Unsafe project id (only [A-Za-z0-9_.:/@-] allowed).");

            bool dropProject = string.IsNullOrEmpty(projectId);
            var outList = new List<string>();
            foreach (var token in Tokenize(template))
            {
                if (token == "{cli}")
                {
                    outList.Add(cliPath);
                }
                else if (token.Contains("{project}") && dropProject)
                {
                    // Drop the placeholder; if it's a bare token preceded by a flag, drop the flag too.
                    if (token == "{project}" && outList.Count > 0
                        && outList[outList.Count - 1].StartsWith("-") && !outList[outList.Count - 1].Contains("="))
                    {
                        outList.RemoveAt(outList.Count - 1);
                    }
                }
                else
                {
                    outList.Add(token.Replace("{env}", env).Replace("{project}", projectId));
                }
            }
            return outList.Where(t => !string.IsNullOrWhiteSpace(t)).ToList();
        }

        /// <summary>Quote-aware tokenizer honoring single and double quotes.</summary>
        public static List<string> Tokenize(string input)
        {
            var tokens = new List<string>();
            var sb = new StringBuilder();
            char? quote = null;
            bool has = false;
            foreach (char c in input)
            {
                if (quote != null)
                {
                    if (c == quote) quote = null; else sb.Append(c);
                }
                else if (c == '"' || c == '\'')
                {
                    quote = c; has = true;
                }
                else if (char.IsWhiteSpace(c))
                {
                    if (has || sb.Length > 0) { tokens.Add(sb.ToString()); sb.Clear(); has = false; }
                }
                else
                {
                    sb.Append(c); has = true;
                }
            }
            if (has || sb.Length > 0) tokens.Add(sb.ToString());
            return tokens;
        }

        /// <summary>Parses the CLI JSON at <paramref name="jsonPath"/>; supports {KEY:VALUE} objects and [{key,value}] arrays.</summary>
        public static IReadOnlyDictionary<string, string> Parse(string output, string jsonPath)
        {
            var secrets = new Dictionary<string, string>();
            if (string.IsNullOrWhiteSpace(output)) return secrets;

            JToken? node;
            try { node = JToken.Parse(output); }
            catch (Exception e) { throw new FetchException($"Failed to parse CLI JSON output: {e.Message}"); }

            if (!string.IsNullOrWhiteSpace(jsonPath) && jsonPath != "$")
            {
                foreach (var part in jsonPath.TrimStart('$').Split('.').Where(p => p.Length > 0))
                {
                    node = node?[part];
                }
            }

            if (node is JArray arr)
            {
                foreach (var item in arr)
                {
                    string? k = (string?)(item["key"] ?? item["secretKey"] ?? item["secretName"]);
                    string? v = (string?)(item["value"] ?? item["secretValue"]);
                    if (k != null && v != null) secrets[k] = v;
                }
            }
            else if (node is JObject obj)
            {
                foreach (var prop in obj.Properties())
                {
                    secrets[prop.Name] = prop.Value?.ToString() ?? "";
                }
            }
            return secrets;
        }

        /// <summary>Windows command-line argument quoting (CommandLineToArgvW rules).</summary>
        private static string QuoteArg(string arg)
        {
            if (arg.Length > 0 && arg.All(c => !char.IsWhiteSpace(c) && c != '"')) return arg;

            var sb = new StringBuilder();
            sb.Append('"');
            int backslashes = 0;
            foreach (char c in arg)
            {
                if (c == '\\')
                {
                    backslashes++;
                }
                else if (c == '"')
                {
                    sb.Append('\\', backslashes * 2 + 1);
                    sb.Append('"');
                    backslashes = 0;
                }
                else
                {
                    if (backslashes > 0) { sb.Append('\\', backslashes); backslashes = 0; }
                    sb.Append(c);
                }
            }
            sb.Append('\\', backslashes * 2);
            sb.Append('"');
            return sb.ToString();
        }
    }
}

File: vscode/src/secretFetcher.ts
/*
 * SecretLoader
 * Copyright (C) 2026 Kivi A.Ş.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { spawn } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { SecretLoaderSettings } from './settings';

export class FetchError extends Error {}

/** A project returned by the project-list command: an id to inject and a human label to show. */
export interface ProjectEntry {
  id: string;
  label: string;
}

// Reject whitespace and anything that could inject extra CLI arguments.
const SAFE_ARG = /^[A-Za-z0-9_.:/@-]+$/;

export async function fetchSecrets(
  s: SecretLoaderSettings,
  env: string,
  projectId: string,
  cwd: string,
): Promise<Record<string, string>> {
  const argv = buildCommand(s.cliPath, s.commandTemplate, env, projectId);
  if (argv.length === 0) throw new FetchError('Command is empty after substitution.');
  const stdout = await run(argv, cwd, s.timeoutSeconds);
  return parse(stdout, s.jsonPath);
}

export async function listProjects(s: SecretLoaderSettings, env: string, cwd: string): Promise<ProjectEntry[]> {
  const argv = buildListCommand(s.cliPath, s.listProjectsCommand, env);
  if (argv.length === 0) throw new FetchError('No project-list command is configured.');
  const stdout = await run(argv, cwd, s.timeoutSeconds);
  return parseProjects(stdout);
}

/** Runs the configured environment-list command and returns the environment slugs. */
export async function listEnvironments(s: SecretLoaderSettings, projectId: string, cwd: string): Promise<string[]> {
  if (!s.listEnvironmentsCommand || !s.listEnvironmentsCommand.trim()) return [];
  // Reuse buildCommand with an empty {env} so {project} substitution + injection guards still apply.
  const argv = buildCommand(s.cliPath, s.listEnvironmentsCommand, '', projectId);
  if (argv.length === 0) return [];
  const stdout = await run(argv, cwd, s.timeoutSeconds);
  return parseEnvironments(stdout);
}

export function buildCommand(cliPath: string, template: string, env: string, projectId: string): string[] {
  if (env.length > 0 && !SAFE_ARG.test(env)) {
    throw new FetchError(`Unsafe environment name '${env}' (only [A-Za-z0-9_.:/@-] allowed).`);
  }
  if (projectId.length > 0 && !SAFE_ARG.test(projectId)) {
    throw new FetchError('Unsafe project id (only [A-Za-z0-9_.:/@-] allowed).');
  }

  const dropProject = projectId.length === 0;
  const out: string[] = [];
  for (const token of tokenize(template)) {
    if (token === '{cli}') {
      out.push(cliPath);
    } else if (token.includes('{project}') && dropProject) {
      // Drop the placeholder; if it's a bare token preceded by a flag, drop the flag too.
      const prev = out[out.length - 1];
      if (token === '{project}' && prev && prev.startsWith('-') && !prev.includes('=')) {
        out.pop();
      }
    } else {
      out.push(token.split('{env}').join(env).split('{project}').join(projectId));
    }
  }
  return out.filter((t) => t.trim().length > 0);
}

export function buildListCommand(cliPath: string, template: string, env: string): string[] {
  if (!template || !template.trim()) return [];
  if (env.length > 0 && !SAFE_ARG.test(env)) {
    throw new FetchError(`Unsafe environment name '${env}' (only [A-Za-z0-9_.:/@-] allowed).`);
  }
  const out: string[] = [];
  for (const token of tokenize(template)) {
    out.push(token === '{cli}' ? cliPath : token.split('{env}').join(env));
  }
  return out.filter((t) => t.trim().length > 0);
}

/** Quote-aware tokenizer honoring single and double quotes. */
export function tokenize(input: string): string[] {
  const tokens: string[] = [];
  let buf = '';
  let quote: string | null = null;
  let has = false;
  for (const c of input) {
    if (quote !== null) {
      if (c === quote) quote = null;
      else buf += c;
    } else if (c === '"' || c === "'") {
      quote = c;
      has = true;
    } else if (/\s/.test(c)) {
      if (has || buf.length > 0) {
        tokens.push(buf);
        buf = '';
        has = false;
      }
    } else {
      buf += c;
      has = true;
    }
  }
  if (has || buf.length > 0) tokens.push(buf);
  return tokens;
}

/** Parses the CLI JSON at jsonPath; supports {KEY:VALUE} objects and [{key,value}] arrays. */
export function parse(output: string, jsonPath: string): Record<string, string> {
  const secrets: Record<string, string> = {};
  if (!output || !output.trim()) return secrets;

  let node: any;
  try {
    node = JSON.parse(output);
  } catch (e) {
    throw new FetchError(`Failed to parse CLI JSON output: ${(e as Error).message}`);
  }

  if (jsonPath && jsonPath.trim() && jsonPath !== '$') {
    for (const part of jsonPath.replace(/^\$/, '').split('.').filter((p) => p.length > 0)) {
      node = node?.[part];
    }
  }

  if (Array.isArray(node)) {
    for (const item of node) {
      const k = item?.key ?? item?.secretKey ?? item?.secretName;
      const v = item?.value ?? item?.secretValue;
      if (typeof k === 'string' && v !== undefined && v !== null) secrets[k] = String(v);
    }
  } else if (node && typeof node === 'object') {
    for (const [k, v] of Object.entries(node)) {
      secrets[k] = v === null || v === undefined ? '' : String(v);
    }
  }
  return secrets;
}

/** Parses a project-list JSON: array of strings, or array of objects (id/name/slug…), under root or .projects/.workspaces/.data. */
export function parseProjects(output: string): ProjectEntry[] {
  const list: ProjectEntry[] = [];
  if (!output || !output.trim()) return list;

  let node: any;
  try {
    node = JSON.parse(output);
  } catch (e) {
    throw new FetchError(`Failed to parse project-list JSON: ${(e as Error).message}`);
  }

  const arr: any[] | undefined = Array.isArray(node)
    ? node
    : node?.projects ?? node?.workspaces ?? node?.data;
  if (!Array.isArray(arr)) return list;

  for (const item of arr) {
    let id: string | undefined;
    let label: string | undefined;
    if (typeof item === 'string') {
      id = label = item;
    } else if (item && typeof item === 'object') {
      id = item.id ?? item.projectId ?? item.workspaceId ?? item.slug ?? item._id;
      label = item.name ?? item.projectName ?? item.slug ?? id;
    }
    if (id) list.push({ id, label: label ?? id });
  }
  return list;
}

/** Parses an environment-list JSON: array of strings, or array of objects (slug/name/id), under root or .environments/.envs/.data. */
export function parseEnvironments(output: string): string[] {
  if (!output || !output.trim()) return [];
  let node: any;
  try {
    node = JSON.parse(output);
  } catch {
    return [];
  }
  const arr: any[] | undefined = Array.isArray(node)
    ? node
    : node?.environments ?? node?.envs ?? node?.data;
  if (!Array.isArray(arr)) return [];

  const out: string[] = [];
  for (const item of arr) {
    if (typeof item === 'string') {
      if (item.trim()) out.push(item.trim());
    } else if (item && typeof item === 'object') {
      const v = item.slug ?? item.name ?? item.id; // slug is what the export command uses (--env=<slug>)
      if (typeof v === 'string' && v.trim()) out.push(v.trim());
    }
  }
  return out;
}

/** Runs argv (no shell — injection-safe), enforces the timeout, checks the exit code, returns stdout. */
function run(argv: string[], cwd: string, timeoutSeconds: number): Promise<string> {
  return new Promise((resolve, reject) => {
    const file = resolveExecutable(argv[0]);
    const rest = argv.slice(1);

    // .cmd/.bat can't be launched directly on Windows — route through cmd.exe, still passing our
    // already-tokenized args as a separate argv (no shell parsing of the tokens).
    const viaCmd = process.platform === 'win32' && /\.(cmd|bat)$/i.test(file);
    const child = viaCmd
      ? spawn('cmd.exe', ['/c', file, ...rest], { cwd, windowsHide: true })
      : spawn(file, rest, { cwd, windowsHide: true });

    let stdout = '';
    let stderr = '';
    let done = false;
    const finish = (fn: () => void) => {
      if (done) return;
      done = true;
      clearTimeout(timer);
      fn();
    };

    const timer = setTimeout(() => {
      try {
        child.kill();
      } catch {
        /* ignore */
      }
      finish(() => reject(new FetchError(`CLI timed out after ${timeoutSeconds}s.`)));
    }, timeoutSeconds * 1000);

    child.stdout?.on('data', (d) => (stdout += d.toString()));
    child.stderr?.on('data', (d) => (stderr += d.toString()));
    child.on('error', (e) => finish(() => reject(new FetchError(`Could not start '${argv[0]}': ${e.message}`))));
    child.on('close', (code) => {
      finish(() => {
        if (code !== 0) {
          let err = stderr.trim();
          if (err.length > 500) err = err.substring(0, 500);
          reject(new FetchError(`CLI exited ${code}: ${err}`));
        } else {
          resolve(stdout);
        }
      });
    });
  });
}

/**
 * Resolves a bare command to a concrete file. POSIX spawn already searches PATH, so this only matters on
 * Windows, where spawn won't append PATHEXT (.exe/.cmd) for a bare name.
 */
function resolveExecutable(cli: string): string {
  if (process.platform !== 'win32') return cli;

  const hasDir = cli.includes('\\') || cli.includes('/');
  if (path.isAbsolute(cli) || hasDir) return withExt(cli) ?? cli;

  for (const dir of (process.env.PATH ?? '').split(path.delimiter).filter(Boolean)) {
    const hit = withExt(path.join(dir, cli));
    if (hit) return hit;
  }
  return cli;
}

function withExt(base: string): string | undefined {
  if (isFile(base)) return base;
  const exts = (process.env.PATHEXT ?? '.COM;.EXE;.BAT;.CMD').split(';').filter(Boolean);
  for (const e of exts) {
    if (isFile(base + e)) return base + e;
    if (isFile(base + e.toLowerCase())) return base + e.toLowerCase();
  }
  return undefined;
}

function isFile(p: string): boolean {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}
